XDT transform causes xml file to have extra data in it (unusable until manually fixed)

Moving bug report to correct area (cross post).
Originally from: https://github.com/chocolatey/simple-server/issues/19

Error occurs during install of [Chocolatey Server (Simple)](https://chocolatey.org/packages/chocolatey.server)

> It looks like a bug in Chocolatey actually - a double transform, where the first corrupts the file and the second can't make changes:
> 
> Preparing transform for 'C:\ProgramData\chocolatey\lib\chocolatey.server\tools\chocolatey.server\Web.config.install.xdt'
> Transforming 'Web.config' with the data from 'Web.config.install.xdt'
> Transform applied successfully for 'C:\ProgramData\chocolatey\lib\chocolatey.server\tools\chocolatey.server\Web.config'
> 
> Preparing transform for 'C:\ProgramData\chocolatey\lib\chocolatey.server\tools\chocolatey.server\bin\Web.config.install.xdt'
> Transforming 'Web.config' with the data from 'Web.config.install.xdt'
> [ERROR] - Error transforming config file:
>  Data at the root level is invalid. Line 230, position 17.

--- a/src/chocolatey/infrastructure.app/services/ConfigTransformService.cs
+++ b/src/chocolatey/infrastructure.app/services/ConfigTransformService.cs
@@ -24,11 +24,13 @@ namespace chocolatey.infrastructure.app.services
     using configuration;
     using filesystem;
     using results;
+    using synchronization;
     using tolerance;
 
     public class ConfigTransformService : IConfigTransformService
     {
         private readonly IFileSystem _fileSystem;
+        private const int MUTEX_TIMEOUT = 2000;
 
         public ConfigTransformService(IFileSystem fileSystem)
         {
@@ -63,27 +65,30 @@ namespace chocolatey.infrastructure.app.services
 
                 foreach (var targetFile in targetFilesTest.or_empty_list_if_null())
                 {
-                    FaultTolerance.try_catch_with_logging_exception(
+                    GlobalMutex.enter(
                         () =>
-                            {
-                                // if there is a backup, we need to put it back in place
-                                // the user has indicated they are using transforms by putting
-                                // the transform file into the folder, so we will override
-                                // the replacement of the file and instead pull from the last 
-                                // backup and let the transform to its thing instead.
-                                var backupTargetFile = targetFile.Replace(ApplicationParameters.PackagesLocation, ApplicationParameters.PackageBackupLocation);
-                                if (_fileSystem.file_exists(backupTargetFile))
-                                {
-                                    this.Log().Debug(()=> "Restoring backup configuration file for '{0}'.".format_with(targetFile));
-                                    _fileSystem.copy_file(backupTargetFile, targetFile, overwriteExisting: true);
-                                }
-                            },
-                        "Error replacing backup config file",
-                        throwError: false,
-                        logWarningInsteadOfError: true);
+                        {
+                            var backupTargetFile = targetFile.Replace(ApplicationParameters.PackagesLocation, ApplicationParameters.PackageBackupLocation);
 
-                    FaultTolerance.try_catch_with_logging_exception(
-                        () =>
+                            FaultTolerance.try_catch_with_logging_exception(
+                                () =>
+                                {
+                                    // if there is a backup, we need to put it back in place
+                                    // the user has indicated they are using transforms by putting
+                                    // the transform file into the folder, so we will override
+                                    // the replacement of the file and instead pull from the last 
+                                    // backup and let the transform to its thing instead.
+                                    if (_fileSystem.file_exists(backupTargetFile))
+                                    {
+                                        this.Log().Debug(()=> "Restoring backup configuration file for '{0}'.".format_with(targetFile));
+                                        _fileSystem.copy_file(backupTargetFile, targetFile, overwriteExisting: true);
+                                    }
+                                },
+                                "Error replacing backup config file",
+                                throwError: false,
+                                logWarningInsteadOfError: true);
+
+                            try
                             {
                                 this.Log().Info(() => "Transforming '{0}' with the data from '{1}'".format_with(_fileSystem.get_file_name(targetFile), _fileSystem.get_file_name(transformFile)));
 
@@ -96,6 +101,12 @@ namespace chocolatey.infrastructure.app.services
                                             document.Load(inputStream);
                                         }
 
+                                        // before applying the XDT transformation, let's make a
+                                        // backup of the file that should be transformed, in case
+                                        // things don't go correctly
+                                        this.Log().Debug(() => "Creating backup configuration file for '{0}'.".format_with(targetFile));
+                                        _fileSystem.copy_file(targetFile, backupTargetFile, overwriteExisting: true);
+
                                         bool succeeded = transformation.Apply(document);
                                         if (succeeded)
                                         {
@@ -106,18 +117,49 @@ namespace chocolatey.infrastructure.app.services
                                                 memoryStream.Seek(0, SeekOrigin.Begin);
                                                 using (var fileStream = _fileSystem.create_file(targetFile))
                                                 {
+                                                    fileStream.SetLength(0);
                                                     memoryStream.CopyTo(fileStream);
                                                 }
+                                            }
+
+                                            // need to test that the transformed configuration file is valid
+                                            // XML.  We can test this by trying to load it again into an XML document
+                                            try
+                                            {
+                                                this.Log().Debug(() => "Verifying transformed configuration file...");
+                                                document.Load(targetFile);
+                                                this.Log().Debug(() => "Transformed configuration file verified.");
+                                            }
+                                            catch (Exception)
+                                            {
+                                                this.Log().Warn(() => "Transformed configuration file doesn't contain valid XML.  Restoring backup file...");
+                                                _fileSystem.copy_file(backupTargetFile, targetFile, overwriteExisting: true);
+                                                this.Log().Debug(() => "Backup file restored.");
                                             }
                                         }
                                         else
                                         {
+                                            // at this point, there is no need to restore the backup file,
+                                            // as the resulting transform hasn't actually been written to disk.
                                             this.Log().Warn(() => "Transform failed for '{0}'".format_with(targetFile));
                                         }
                                     }
                                 }
-                            },
-                        "Error transforming config file");
+                            }
+                            catch (Exception)
+                            {
+                                FaultTolerance.try_catch_with_logging_exception(
+                                    () =>
+                                    {
+                                        // something went wrong with the transformation, so we should restore
+                                        // the original configuration file from the backup
+                                        this.Log().Warn(() => "There was a problem transforming configuration file, restoring backup file...");
+                                        _fileSystem.copy_file(backupTargetFile, targetFile, overwriteExisting: true);
+                                        this.Log().Debug(() => "Backup file restored.");
+                                    },
+                                    "Error restoring backup configuration file.");
+                            }
+                        }, MUTEX_TIMEOUT);
                 }
             }
         }

